azurelinuxagent/common/persist_firewall_rules.py

```diff
@@ -0,0 +1,235 @@
+# Do not edit.
+[Service]
+# The first line clears the old data, and the 2nd line overwrites it.
+Environment=
+Environment="EGG={egg_path}" "DST_IP={wire_ip}" "UID={user_id}" "WAIT={wait}"
+"""
+
+    _AGENT_NETWORK_SETUP_NAME_FORMAT = "{0}-network-setup.service"
+    _DROP_IN_ENV_FILE_NAME = "10-environment.conf"
+
+    _FIREWALLD_RUNNING_CMD = ["firewall-cmd", "--state"]
+
+    def __init__(self, dst_ip, uid):
+        """
+        This class deals with ensuring that Firewall rules are persisted over system reboots.
+        It tries to employ using Firewalld.service if present first as it already has provisions for persistent rules.
+        If not, it then creates a new agent-network-setup.service file and copy it over to the osutil.get_systemd_unit_file_install_path() dynamically
+        On top of it, on every service restart it ensures that the WireIP is overwritten and the new IP is blocked as well.
+        """
+        osutil = get_osutil()
+        self._network_setup_service_name = self._AGENT_NETWORK_SETUP_NAME_FORMAT.format(osutil.get_service_name())
+        self._is_systemd = CGroupsApi.is_systemd()
+        self._systemd_file_path = osutil.get_systemd_unit_file_install_path()
+        self._dst_ip = dst_ip
+        self._uid = uid
+        self._wait = osutil.get_firewall_will_wait()
+        # The custom service will try to call the current agent executable to setup the firewall rules
+        self._current_agent_executable_path = os.path.join(os.getcwd(), sys.argv[0])
+
+    @staticmethod
+    def _is_firewall_service_running():
+        # Check if firewall-cmd can connect to the daemon
+        # https://docs.fedoraproject.org/en-US/Fedora/19/html/Security_Guide/sec-Check_if_firewalld_is_running.html
+        # Eg:    firewall-cmd --state
+        #           >   running
+        firewalld_state = PersistFirewallRulesHandler._FIREWALLD_RUNNING_CMD
+        try:
+            return shellutil.run_command(firewalld_state).rstrip() == "running"
+        except Exception as error:
+            logger.info("{0} command returned error/not running: {1}".format(' '.join(firewalld_state), ustr(error)))
+        return False
+
+    def setup(self):
+        if self._is_firewall_service_running():
+            logger.info("Firewalld.service present on the VM, setting up permanent rules on the VM")
+            # In case of a failure, this would throw. In such a case, we don't need to try to setup our custom service
+            # because on system reboot, all iptable rules are reset by firewalld.service so it would be a no-op.
+            self._setup_permanent_firewalld_rules()
+            return
+
+        logger.info(
+            "Firewalld service not running/unavailable, trying to set up {0}".format(self._network_setup_service_name))
+
+        if not CGroupsApi.is_systemd():
+            raise Exception("Systemd not enabled, unable to set {0}".format(self._network_setup_service_name))
```

Comment:
Let's rephrase ""Systemd not enabled" to "Did not detect systemd" or something similar

Reply:
Done